runtime/cudaq/dynamics/helpers.cpp

```diff
@@ -0,0 +1,92 @@
+/*******************************************************************************
+ * Copyright (c) 2022 - 2025 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
+#include <unordered_map>
+#include "helpers.h"
+
+namespace cudaq {
+namespace detail {
+
+  std::vector<std::string>
+  generate_all_states(const std::vector<int> &degrees, const std::unordered_map<int, int> &dimensions) {
+    if (degrees.size() == 0)
+      return {};
+
+    std::vector<std::string> states;
+    auto entry = dimensions.find(degrees[0]);
+    assert (entry != dimensions.end());
+    for (auto state = 0; state < entry->second; state++) {
+      states.push_back(std::to_string(state));
+    }
+
+    for (auto idx = 1; idx < degrees.size(); ++idx) {
+      auto entry = dimensions.find(degrees[idx]);
+      assert (entry != dimensions.end());
+      std::vector<std::string> result;
+      for (auto current : states) {
+        for (auto state = 0; state < entry->second; state++) {
+          result.push_back(current + std::to_string(state));
+        }
+      }
+      states = result;
+    }
+
+    return states;
+  }
+
+  std::vector<int> compute_permutation(const std::vector<int> &op_degrees,
+                                       const std::vector<int> &canon_degrees, 
+                                       const std::unordered_map<int, int> dimensions) {
+    assert(op_degrees.size() == canon_degrees.size());
+    auto states = cudaq::detail::generate_all_states(canon_degrees, dimensions);
+  
+    std::vector<int> reordering;
+    for (auto degree : op_degrees) {
+      auto it = std::find(canon_degrees.cbegin(), canon_degrees.cend(), degree);
+      reordering.push_back(it - canon_degrees.cbegin());
+    }
+  
+    std::vector<std::string> op_states = 
+        cudaq::detail::generate_all_states(op_degrees, dimensions);
+  
+    std::vector<int> permutation;
+    for (auto state : states) {
```

Comment:
nit: we can save a string copy here
```suggestion
    for (const auto& state : states) {
```